allennlp/modules/elmo.py

```diff
@@ -0,0 +1,422 @@
+        """
+        bilm_output = self._elmo_lstm(inputs)
+        layer_activations = bilm_output['activations']
+        mask_with_bos_eos = bilm_output['mask']
+
+        elmo_representations = []
+        for scalar_mix in self._scalar_mixes:
+            representation_with_bos_eos = scalar_mix.forward(layer_activations, mask_with_bos_eos)
+            representation_without_bos_eos, mask_without_bos_eos = remove_sentence_boundaries(
+                    representation_with_bos_eos, mask_with_bos_eos
+            )
+            elmo_representations.append(representation_without_bos_eos)
+
+        return {'elmo': elmo_representations, 'mask': mask_without_bos_eos}
+
+    @classmethod
+    def from_params(cls, params: Params) -> 'Elmo':
+        options_file = params.pop('options_file')
+        weight_file = params.pop('weight_file')
+        num_elmo_layers = params.pop('num_elmo_layers')
+        do_layer_norm = params.pop('do_layer_norm', False)
+        params.assert_empty(cls.__name__)
+        return cls(options_file, weight_file, num_elmo_layers, do_layer_norm)
+
+
+class _ElmoTokenRepresentation(torch.nn.Module):
+    """
+    Compute context sensitive token representation using pretrained biLM.
+
+    This embedder has input character ids of size (batch_size, sequence_length, 50)
+    and returns (batch_size, sequence_length + 2, embedding_dim), where embedding_dim
+    is specified in the options file (typically 512).
+
+    We add special entries at the beginning and end of each sequence corresponding
+    to <S> and </S>, the beginning and end of sentence tokens.
+
+    Note: this is a lower level class useful for advanced usage.  Most users should
+    use ``ElmoTokenEmbedder`` or ``allennlp.modules.Elmo`` instead.
+
+    Parameters
+    ----------
+    options_file : str
+        ELMo JSON options file
+    weight_file : str
+        ELMo hdf5 weight file
+
+    The relevant section of the options file is something like:
+    .. example-code::
+
+        .. code-block:: python
+
+            {'char_cnn': {
+                'activation': 'relu',
+                'embedding': {'dim': 4},
+                'filters': [[1, 4], [2, 8], [3, 16], [4, 32], [5, 64]],
+                'max_characters_per_token': 50,
+                'n_characters': 262,
+                'n_highway': 2
+                }
+            }
+
+    something
+    """
+    def __init__(self,
+                 options_file: str,
+                 weight_file: str) -> None:
+        super(_ElmoTokenRepresentation, self).__init__()
+
+        with open(cached_path(options_file), 'r') as fin:
+            self._options = json.load(fin)
+        self._weight_file = weight_file
+
+        self.output_dim = self._options['lstm']['projection_dim']
+
+        self._load_weights()
+
+        # Cache the arrays for use in forward -- +1 due to masking.
+        self._beginning_of_sentence_characters = Variable(torch.from_numpy(
```

Comment:
This seems bad to mutate the id of this here. Why doesn't `ELMoCharacterMapper` take this into account in the first place?

Reply:
Agree this breaks encapsulation.  TODO to clean up later.  ELMoCharacterMapper doesn't know anything about masking/padding as that happens in the indexer. So we need to add +1 here.